lib/chef/provider/package/apt.rb

```diff
@@ -50,86 +56,91 @@ def default_release_options
+            case line
+            when /^\s{2}Installed: (.+)$/
+              installed_version = $1
+              if installed_version == '(none)'
+                Chef::Log.debug("#{@new_resource} current version is nil")
+                installed_version = nil
+              else
+                Chef::Log.debug("#{@new_resource} current version is #{installed_version}")
+                installed = true
+              end
+            when /^\s{2}Candidate: (.+)$/
+              candidate_version = $1
+              if candidate_version == '(none)'
+                # This may not be an appropriate assumption, but it shouldn't break anything that already worked -- btm
+                is_virtual_package = true
+                showpkg = shell_out!("apt-cache showpkg #{pkg}", {:timeout => 900}).stdout
+                providers = Hash.new
+                showpkg.rpartition(/Reverse Provides: ?#{$/}/)[2].each_line do |line|
+                  provider, version = line.split
+                  providers[provider] = version
                 end
+                # Check if the package providing this virtual package is installed
+                num_providers = providers.length
+                raise Chef::Exceptions::Package, "#{@new_resource.package_name} has no candidate in the apt-cache" if num_providers == 0
+                # apt will only install a virtual package if there is a single providing package
+                raise Chef::Exceptions::Package, "#{@new_resource.package_name} is a virtual package provided by #{num_providers} packages, you must explicitly select one to install" if num_providers > 1
+                # Check if the package providing this virtual package is installed
+                Chef::Log.info("#{@new_resource} is a virtual package, actually acting on package[#{providers.keys.first}]")
+                ret = check_package_state(providers.keys.first)
+                installed = ret[:installed]
+                installed_version = ret[:installed_version]
+              else
+                Chef::Log.debug("#{@new_resource} candidate version is #{$1}")
               end
             end
-            if package.is_a?(Array)
-              final_installed_version << installed_version
-              final_candidate_version << candidate_version
-              final_installed << installed
-              final_virtual << virtual
-            else
-              final_installed_version = installed_version
-              final_candidate_version = candidate_version
-              final_installed = installed
-              final_virtual = virtual
-            end
           end
-          @candidate_version = final_candidate_version
-          @current_resource.version(final_installed_version)
-          @is_virtual_package = final_virtual
- 
-          return final_installed.is_a?(Array) ? final_installed.any? : final_installed
+
+          return {
+            installed_version:   installed_version,
```

Comment:
I :heart: it so much.

I feel like I'm giving the middle finger to ruby 1.8.7 every time I use it.  It fills me with great joy =)